lib/ansible/modules/network/panos/panos_address.py

```diff
@@ -0,0 +1,215 @@
+
+RETURN = '''
+# Default return values
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.basic import get_exception
+
+try:
+    import pan.xapi
+    from pan.xapi import PanXapiError
+
+    HAS_LIB = True
+except ImportError:
+    HAS_LIB = False
+
+_ADDRESS_XPATH = "/config/devices/entry[@name='localhost.localdomain']" + \
+                 "/vsys/entry[@name='vsys1']" + \
+                 "/address/entry[@name='%s']"
+
+
+def address_exists(xapi, address_name):
+    xapi.get(_ADDRESS_XPATH % address_name)
+    e = xapi.element_root.find('.//entry')
+    if e is None:
+        return False
+    return True
+
+
+def add_address(xapi, module, address, address_name, description, type, tag):
+    if address_exists(xapi, address_name):
+        return False
+
+    exml = []
+    exml.append('<%s>' % type)
+    exml.append('%s' % address)
+    exml.append('</%s>' % type)
+
+    if description:
+        exml.append('<description>')
+        exml.append('%s' % description)
+        exml.append('</description>')
+
+    if tag:
+        exml.append('<tag>')
+        exml.append('<member>%s</member>' % tag)
+        exml.append('</tag>')
+
+    exml = ''.join(exml)
+
+    xapi.set(xpath=_ADDRESS_XPATH % address_name, element=exml)
+
+    return True
+
+
+def main():
+    argument_spec = dict(
+        ip_address=dict(required=True),
+        password=dict(required=True, no_log=True),
+        username=dict(default='admin'),
+        address_name=dict(required=True),
+        address=dict(default=None),
+        description=dict(default=None),
+        tag=dict(default=None),
+        type=dict(default='ip-netmask', choices=['ip-netmask', 'ip-range', 'fqdn']),
+        commit=dict(type='bool', default=True)
+    )
+    module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=False)
+
+    if not HAS_LIB:
+        module.fail_json(msg='pan-python required for this module')
+
+    ip_address = module.params["ip_address"]
+    password = module.params["password"]
+    username = module.params['username']
+
+    xapi = pan.xapi.PanXapi(
+        hostname=ip_address,
+        api_username=username,
+        api_password=password
+    )
+
+    address_name = module.params['address_name']
+    address = module.params['address']
+    commit = module.params['commit']
+
+    description = module.params['description']
+    tag = module.params['tag']
+    type = module.params['type']
+
+    changed = False
+    try:
+        changed = add_address(xapi, module,
+                              address,
+                              address_name,
+                              description,
+                              type,
+                              tag)
+    except PanXapiError:
+        exc = get_exception()
+        module.fail_json(msg=exc.message)
+
+    if changed and commit:
+        xapi.commit(cmd="<commit></commit>", sync=True, interval=1)
+
+    module.exit_json(changed=changed, msg="okey dokey")
```

Comment:
Not really, the call does not return anything. We are using this message for past 2 years and it is a bit of 'developer justice.' Customers seem to have no issue with it.